From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 15 Apr 2026 23:58:48 -0400
Subject: [PATCH] Improve Purpur JS expression evaluation

Fallback to vanilla logic if the related Purpur config value is default, to skip unnecessary JS evaluation.
And use fastutil's map to prevent the cost of un/boxing.

Actually, we can also replace the Rhino with GraalJS to achieve better performance. However, I don't think
there are many people who will use this Purpur feature, we can do it if really needed.

diff --git a/net/minecraft/world/entity/monster/Phantom.java b/net/minecraft/world/entity/monster/Phantom.java
index 5da6b882a2f42330d7d912dd6edf76ab5f970acb..df9f7ea921601693130ea9d4bd7f6a9438945fa5 100644
--- a/net/minecraft/world/entity/monster/Phantom.java
+++ b/net/minecraft/world/entity/monster/Phantom.java
@@ -182,10 +182,10 @@ public class Phantom extends Mob implements Enemy {
     private void updatePhantomSizeInfo() {
         this.refreshDimensions();
         if (level().purpurConfig.phantomFlamesOnSwoop && attackPhase == AttackPhase.SWOOP) shoot(); // Purpur - Ridables - Phantom flames on swoop
-        // Purpur start - Configurable entity base attributes
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(getFromCache(() -> this.level().purpurConfig.phantomMaxHealth, () -> this.level().purpurConfig.phantomMaxHealthCache, () -> 20.0D));
-        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(getFromCache(() -> this.level().purpurConfig.phantomAttackDamage, () -> this.level().purpurConfig.phantomAttackDamageCache, () -> (double) (6 + this.getPhantomSize())));
-        // Purpur end - Configurable entity base attributes
+        // Leaf start - Improve Purpur JS expression evaluation
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(leaf$getMaxHealth(this));
+        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(leaf$getAttackDamage(this));
+        // Leaf end - Improve Purpur JS expression evaluation
     }
 
     public int getPhantomSize() {
@@ -222,6 +222,24 @@ public class Phantom extends Mob implements Enemy {
     }
     // Purpur end - Configurable entity base attributes
 
+    // Leaf start - Improve Purpur JS expression evaluation
+    private static double leaf$getMaxHealth(final Phantom phantom) {
+        final Level level = phantom.level();
+        if (level.purpurConfig.phantomMaxHealthEnabled) {
+            return phantom.getFromCache(() -> level.purpurConfig.phantomMaxHealth, () -> level.purpurConfig.phantomMaxHealthCache, () -> 20.0D); // Purpur - Configurable entity base attributes
+        }
+        return 20.0D;
+    }
+
+    private static double leaf$getAttackDamage(final Phantom phantom) {
+        final Level level = phantom.level();
+        if (level.purpurConfig.phantomAttackDamageEnabled) {
+            return phantom.getFromCache(() -> level.purpurConfig.phantomAttackDamage, () -> level.purpurConfig.phantomAttackDamageCache, () -> (double) (6 + phantom.getPhantomSize())); // Purpur - Configurable entity base attributes
+        }
+        return 6 + phantom.getPhantomSize();
+    }
+    // Leaf end - Improve Purpur JS expression evaluation
+
     @Override
     public void tick() {
         super.tick();
diff --git a/net/minecraft/world/entity/monster/Shulker.java b/net/minecraft/world/entity/monster/Shulker.java
index 4c578c6890b636eab674bb1f18100b14beb10d10..e135fc4985b8de14523b7cfa4d331c7ee55708ea 100644
--- a/net/minecraft/world/entity/monster/Shulker.java
+++ b/net/minecraft/world/entity/monster/Shulker.java
@@ -536,7 +536,7 @@ public class Shulker extends AbstractGolem implements Enemy {
             if (!this.level().purpurConfig.shulkerSpawnFromBulletNearbyEquation.isBlank()) {
                 int shulkerCount = this.level().getEntities(EntityTypes.SHULKER, oldAabb.inflate(this.level().purpurConfig.shulkerSpawnFromBulletNearbyRange), Entity::isAlive).size();
                 try {
-                    failureChance -= ((Number) scriptEngine.eval("let nearby = " + shulkerCount + "; " + this.level().purpurConfig.shulkerSpawnFromBulletNearbyEquation)).floatValue();
+                    failureChance -= this.level().purpurConfig.shulkerSpawnFromBulletNearbyEquationEnabled ? ((Number) scriptEngine.eval("let nearby = " + shulkerCount + "; " + this.level().purpurConfig.shulkerSpawnFromBulletNearbyEquation)).floatValue() : (shulkerCount - 1) / 5.0F; // Leaf - Improve Purpur JS expression evaluation
                 } catch (javax.script.ScriptException e) {
                     e.printStackTrace();
                     failureChance -= (shulkerCount - 1) / 5.0F;
diff --git a/net/minecraft/world/entity/monster/cubemob/Slime.java b/net/minecraft/world/entity/monster/cubemob/Slime.java
index 2472fce146b9122098e9954757b0687b68122404..0e582eab13bfbfa3e973b8308eb01ba1ba1b00ab 100644
--- a/net/minecraft/world/entity/monster/cubemob/Slime.java
+++ b/net/minecraft/world/entity/monster/cubemob/Slime.java
@@ -212,8 +212,26 @@ public class Slime extends AbstractCubeMob implements Enemy {
     @Override
     protected void setCubeMobHealth(final int actualSize) {
         // Purpur start - Configurable entity base attributes
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(getFromCache(this::getMaxHealthEquation, this::getMaxHealthCache, () -> (double) (actualSize * actualSize)));
-        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(getFromCache(this::getAttackDamageEquation, this::getAttackDamageCache, () -> (double) actualSize));
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(leaf$getMaxHealth(this, actualSize)); // Leaf - Improve Purpur JS expression evaluation
+        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(leaf$getAttackDamage(this, actualSize)); // Leaf - Improve Purpur JS expression evaluation
         // Purpur end - Configurable entity base attributes
     }
+
+    // Leaf start - Improve Purpur JS expression evaluation
+    private static double leaf$getMaxHealth(final Slime slime, final int size) {
+        final Level level = slime.level();
+        if (level.purpurConfig.slimeMaxHealthEnabled) {
+            return slime.getFromCache(slime::getMaxHealthEquation, slime::getMaxHealthCache, () -> (double) (size * size));
+        }
+        return size * size;
+    }
+
+    private static double leaf$getAttackDamage(final Slime slime, final int attackDamage) {
+        final Level level = slime.level();
+        if (level.purpurConfig.slimeAttackDamageEnabled) {
+            return slime.getFromCache(slime::getAttackDamageEquation, slime::getAttackDamageCache, () -> (double) attackDamage);
+        }
+        return attackDamage;
+    }
+    // Leaf end - Improve Purpur JS expression evaluation
 }
diff --git a/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java b/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
index d236658c67feaf0563e47411ceab366a61346523..ea54b1d881abb45a166efb6a632027b1ff6c50b9 100644
--- a/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
@@ -187,7 +187,7 @@ public abstract class AbstractSkeleton extends Monster implements RangedAttackMo
         double distanceToTarget = Math.sqrt(xd * xd + zd * zd);
         if (this.level() instanceof ServerLevel serverLevel) {
             Projectile.Delayed<AbstractArrow> delayedEntity = Projectile.spawnProjectileUsingShootDelayed( // Paper - delayed
-                arrow, serverLevel, projectile, xd, yd + distanceToTarget * 0.2F, zd, 1.6F, serverLevel.purpurConfig.skeletonBowAccuracyMap.getOrDefault(serverLevel.getDifficulty().getId(), this.rangedAttackUncertainty(serverLevel))  // Purpur - skeleton bow accuracy option
+                arrow, serverLevel, projectile, xd, yd + distanceToTarget * 0.2F, zd, 1.6F, this.leaf$getBowAccuracy(serverLevel) // Leaf - Improve Purpur JS expression evaluation
             );
 
             // Paper start - call EntityShootBowEvent
@@ -206,6 +206,15 @@ public abstract class AbstractSkeleton extends Monster implements RangedAttackMo
         this.playSound(SoundEvents.SKELETON_SHOOT, 1.0F, 1.0F / (this.getRandom().nextFloat() * 0.4F + 0.8F));
     }
 
+    // Leaf start - Improve Purpur JS expression evaluation
+    private float leaf$getBowAccuracy(final ServerLevel serverLevel) {
+        if (serverLevel.purpurConfig.skeletonBowAccuracyEnabled) {
+            return serverLevel.purpurConfig.skeletonBowAccuracyMap.getOrDefault(serverLevel.getDifficulty().getId(), this.rangedAttackUncertainty(serverLevel)); // Purpur - skeleton bow accuracy option
+        }
+        return this.rangedAttackUncertainty(serverLevel);
+    }
+    // Leaf end - Improve Purpur JS expression evaluation
+
     protected AbstractArrow getArrow(final ItemStack projectile, final float power, final @Nullable ItemStack firingWeapon) {
         return ProjectileUtil.getMobArrow(this, projectile, power, firingWeapon);
     }
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index d43bae98bff9bc6dcf4108ff4b5f59d1a4435894..0599d3172fba2dad74ac834db36a34a57b496fef 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1809,7 +1809,10 @@ public abstract class Player extends Avatar implements ContainerUser {
     protected int getBaseExperienceReward(final ServerLevel level) {
         // Purpur start - Add player death exp control options
         if (!level.getGameRules().get(GameRules.KEEP_INVENTORY) && !this.isSpectator()) {
-            int toDrop;
+            // Leaf start - Improve Purpur JS expression evaluation
+            int toDrop = experienceLevel * 7;
+            if (level().purpurConfig.playerDeathExpDropEquationEnabled) {
+                // Leaf end - Improve Purpur JS expression evaluation
             try {
                 toDrop = Math.round(((Number) scriptEngine.eval("let expLevel = " + experienceLevel + "; " +
                     "let expTotal = " + totalExperience + "; " +
@@ -1819,6 +1822,7 @@ public abstract class Player extends Avatar implements ContainerUser {
                 e.printStackTrace();
                 toDrop = experienceLevel * 7;
             }
+            } // Leaf - Improve Purpur JS expression evaluation
             return Math.min(toDrop, level().purpurConfig.playerDeathExpDropMax);
         } else {
             return 0;
diff --git a/org/purpurmc/purpur/PurpurWorldConfig.java b/org/purpurmc/purpur/PurpurWorldConfig.java
index 43fb8c658a0532cd48047599daff87a8b57898f2..a04374f534e4a419776b4cabf32c17077e7847e7 100644
--- a/org/purpurmc/purpur/PurpurWorldConfig.java
+++ b/org/purpurmc/purpur/PurpurWorldConfig.java
@@ -473,6 +473,7 @@ public class PurpurWorldConfig {
     public boolean idleTimeoutUpdateTabList = false;
     public boolean idleTimeoutTargetPlayer = true;
     public String playerDeathExpDropEquation = "expLevel * 7";
+    public boolean playerDeathExpDropEquationEnabled; // Leaf - Improve Purpur JS expression evaluation
     public int playerDeathExpDropMax = 100;
     public boolean teleportIfOutsideBorder = false;
     public boolean teleportOnNetherCeilingDamage = false;
@@ -521,6 +522,7 @@ public class PurpurWorldConfig {
         shiftRightClickRepairsMendingPoints = getInt("gameplay-mechanics.player.shift-right-click-repairs-mending-points", shiftRightClickRepairsMendingPoints);
         playerExpPickupDelay = getInt("gameplay-mechanics.player.exp-pickup-delay-ticks", playerExpPickupDelay);
         playerVoidTrading = getBoolean("gameplay-mechanics.player.allow-void-trading", playerVoidTrading);
+        playerDeathExpDropEquationEnabled = !"expLevel * 7".equals(playerDeathExpDropEquation); // Leaf - Improve Purpur JS expression evaluation
     }
 
     public boolean silkTouchEnabled = false;
@@ -2434,6 +2436,10 @@ public class PurpurWorldConfig {
     public boolean phantomAllowGriefing = false;
     public String phantomMaxHealth = "20.0";
     public String phantomAttackDamage = "6 + size";
+    // Leaf start - Improve Purpur JS expression evaluation
+    public boolean phantomMaxHealthEnabled;
+    public boolean phantomAttackDamageEnabled;
+    // Leaf end - Improve Purpur JS expression evaluation
     public Map<Integer, Double> phantomMaxHealthCache = new HashMap<>();
     public Map<Integer, Double> phantomAttackDamageCache = new HashMap<>();
     public double phantomAttackedByCrystalRadius = 0.0D;
@@ -2472,6 +2478,10 @@ public class PurpurWorldConfig {
         }
         phantomMaxHealth = getString("mobs.phantom.attributes.max_health", phantomMaxHealth);
         phantomAttackDamage = getString("mobs.phantom.attributes.attack_damage", phantomAttackDamage);
+        // Leaf start - Improve Purpur JS expression evaluation
+        phantomMaxHealthEnabled = !"20.0".equals(phantomMaxHealth);
+        phantomAttackDamageEnabled = !"6 + size".equals(phantomAttackDamage);
+        // Leaf end - Improve Purpur JS expression evaluation
         phantomMaxHealthCache.clear();
         phantomAttackDamageCache.clear();
         phantomAttackedByCrystalRadius = getDouble("mobs.phantom.attacked-by-crystal-range", phantomAttackedByCrystalRadius);
@@ -2880,6 +2890,7 @@ public class PurpurWorldConfig {
     public boolean shulkerSpawnFromBulletRequireOpenLid = true;
     public double shulkerSpawnFromBulletNearbyRange = 8.0D;
     public String shulkerSpawnFromBulletNearbyEquation = "(nearby - 1) / 5.0";
+    public boolean shulkerSpawnFromBulletNearbyEquationEnabled; // Leaf - Improve Purpur JS expression evaluation
     public boolean shulkerSpawnFromBulletRandomColor = false;
     public boolean shulkerChangeColorWithDye = false;
     public boolean shulkerAlwaysDropExp = false;
@@ -2899,6 +2910,7 @@ public class PurpurWorldConfig {
         shulkerSpawnFromBulletRequireOpenLid = getBoolean("mobs.shulker.spawn-from-bullet.require-open-lid", shulkerSpawnFromBulletRequireOpenLid);
         shulkerSpawnFromBulletNearbyRange = getDouble("mobs.shulker.spawn-from-bullet.nearby-range", shulkerSpawnFromBulletNearbyRange);
         shulkerSpawnFromBulletNearbyEquation = getString("mobs.shulker.spawn-from-bullet.nearby-equation", shulkerSpawnFromBulletNearbyEquation);
+        shulkerSpawnFromBulletNearbyEquationEnabled = !"(nearby - 1) / 5.0".equals(shulkerSpawnFromBulletNearbyEquation);
         shulkerSpawnFromBulletRandomColor = getBoolean("mobs.shulker.spawn-from-bullet.random-color", shulkerSpawnFromBulletRandomColor);
         shulkerChangeColorWithDye = getBoolean("mobs.shulker.change-color-with-dye", shulkerChangeColorWithDye);
         shulkerAlwaysDropExp = getBoolean("mobs.shulker.always-drop-exp", shulkerAlwaysDropExp);
@@ -2946,7 +2958,8 @@ public class PurpurWorldConfig {
     public boolean skeletonAlwaysDropExp = false;
     public int skeletonFeedWitherRoses = 0;
     public String skeletonBowAccuracy = "14 - difficulty * 4";
-    public Map<Integer, Float> skeletonBowAccuracyMap = new HashMap<>();
+    public boolean skeletonBowAccuracyEnabled; // Leaf - Improve Purpur JS expression evaluation
+    public it.unimi.dsi.fastutil.ints.Int2FloatArrayMap skeletonBowAccuracyMap = new it.unimi.dsi.fastutil.ints.Int2FloatArrayMap(); // Leaf - Improve Purpur JS expression evaluation
     public Boolean skeletonCanPickUpLoot = null;
     private void skeletonSettings() {
         skeletonRidable = getBoolean("mobs.skeleton.ridable", skeletonRidable);
@@ -2964,6 +2977,7 @@ public class PurpurWorldConfig {
         skeletonFeedWitherRoses = getInt("mobs.skeleton.feed-wither-roses", skeletonFeedWitherRoses);
         final String defaultSkeletonBowAccuracy = skeletonBowAccuracy;
         skeletonBowAccuracy = getString("mobs.skeleton.bow-accuracy", skeletonBowAccuracy);
+        skeletonBowAccuracyEnabled = !"14 - difficulty * 4".equals(skeletonBowAccuracy); // Leaf - Improve Purpur JS expression evaluation
         for (int i = 1; i < 4; i++) {
             final float divergence;
             try {
@@ -3013,8 +3027,12 @@ public class PurpurWorldConfig {
     public boolean slimeControllable = true;
     public String slimeMaxHealth = "size * size";
     public String slimeAttackDamage = "size";
-    public Map<Integer, Double> slimeMaxHealthCache = new HashMap<>();
-    public Map<Integer, Double> slimeAttackDamageCache = new HashMap<>();
+    // Leaf start - Improve Purpur JS expression evaluation
+    public boolean slimeMaxHealthEnabled;
+    public boolean slimeAttackDamageEnabled;
+    public it.unimi.dsi.fastutil.ints.Int2DoubleArrayMap slimeMaxHealthCache = new it.unimi.dsi.fastutil.ints.Int2DoubleArrayMap();
+    public it.unimi.dsi.fastutil.ints.Int2DoubleArrayMap slimeAttackDamageCache = new it.unimi.dsi.fastutil.ints.Int2DoubleArrayMap();
+    // Leaf end - Improve Purpur JS expression evaluation
     public boolean slimeTakeDamageFromWater = false;
     public boolean slimeAlwaysDropExp = false;
     private void slimeSettings() {
@@ -3028,6 +3046,10 @@ public class PurpurWorldConfig {
         }
         slimeMaxHealth = getString("mobs.slime.attributes.max_health", slimeMaxHealth);
         slimeAttackDamage = getString("mobs.slime.attributes.attack_damage", slimeAttackDamage);
+        // Leaf start - Improve Purpur JS expression evaluation
+        slimeMaxHealthEnabled = !"size * size".equals(slimeMaxHealth);
+        slimeAttackDamageEnabled = !"size".equals(slimeAttackDamage);
+        // Leaf end - Improve Purpur JS expression evaluation
         slimeMaxHealthCache.clear();
         slimeAttackDamageCache.clear();
         slimeTakeDamageFromWater = getBoolean("mobs.slime.takes-damage-from-water", slimeTakeDamageFromWater);
